src/CpuAffinity.cc

```diff
@@ -64,3 +67,21 @@ CpuAffinityCheck()
     }
 }
 
+class CpuAffinityRr : public RegisteredRunner
+{
+public:
+    virtual void finalizeConfig() override {
```

Comment:
The finalizeConfig() event has a specific and documented purpose that calling CpuAffinityInit() does not match. We should use a different event for CpuAffinityInit(). I will sketch two mutually exclusive solutions:

1. If CpuAffinityInit() does not require enter_suid() privileges, then it should be called from useConfig():

```suggestion
    virtual void useConfig() override {
```

2. If CpuAffinityInit() requires privileges that useConfig() does not provide, then we need to add a RegisteredRunner event dedicated to that special need and use that here. For example:

```suggestion
    virtual void useConfigWithPriviliges() override {
```

The useConfigWithPriviliges() event can be triggered where CpuAffinityInit() is called now AFAICT.

----

AFAICT, the action in question here is sched_setaffinity(). On Linux, that action requires:

* an effective user ID equal to the real user ID;
* an effective user ID equal to the effective user ID of the thread identified by pid; or
* possession of  the CAP_SYS_NICE capability in the user namespace of the thread pid.

AFAICT, the official code always runs sched_setaffinity() with enter_suid() privileges. Most likely we must continue to do so. In that case, my second suggestion applies.

Reply:
> The finalizeConfig() event has a specific and documented purpose that calling CpuAffinityInit() does not match.

IMO, affinity is currently lacking checks that the CPUs match process numbers. That depends on configuration for workers and cache_dir etc.

Also, (performance benefit) affinity assigned before claimMemoryNeeds() allows the OS to ring-fence the memory allocations to the CPU assigned that process.


> AFAICT, the action in question here is sched_setaffinity(). On Linux, that action requires:
> 
>     * an effective user ID equal to the real user ID;
> 
>     * an effective user ID equal to the effective user ID of the thread identified by pid; or
> 
>     * possession of  the CAP_SYS_NICE capability in the user namespace of the thread pid.
> 
> 
> AFAICT, the official code always runs sched_setaffinity() with enter_suid() privileges. Most likely we must continue to do so. In that case, my second suggestion applies.

Thank you for that info. The code running with high privileges should be minimized, which means each runner actually needing it should do the calls itself.

Reply:
> > Alex: The finalizeConfig() event has a specific and documented purpose that calling CpuAffinityInit() does not match.

> IMO, affinity is currently lacking checks that the CPUs match process numbers. That depends on configuration for workers and cache_dir etc.

> Also, (performance benefit) affinity assigned before claimMemoryNeeds() allows the OS to ring-fence the memory allocations to the CPU assigned that process.

The mismatch I was talking about does not depend on the things you are talking about. If you cannot find a matching event, let's add a new (matching) event, but please do not put this code into a mismatching event handler. The finalizeConfig() is a mismatching event.

Assuming you can justify adding a new event, please do not add that new event _before_ claimMemoryNeeds() because doing so may prevent future code handling the new event from allocating certain memory blocks. Please note that claimMemoryNeeds() does _not_ allocate memory; large memory allocations should start _after_ claimMemoryNeeds() completes.

> The code running with high privileges should be minimized, 

Agreed.

> which means each runner actually needing it should do the calls itself.

I have not studied this issue; I am not against this "enter/leave_suid() as needed" approach if you are sure it is the best way forward despite how much each enter_suid() and leave_suid() calls do already.

Please note that both enter_suid() and leave_suid() depend on squid.conf and, hence, they should not be called until _after_ the finalizeConfig() event completes (at least).

Reply:
CpuAffinityCheck() which is called from here depends on all workers and cache_dir in order to validate the cpu_affinity_map directive value. IMO that meets the criteria for finalizeConfig().

Reply:
Oh. I see what you mean.

That means we have a documentation issue with finalizeConfig(). It clearly states that it is for ```options that depend on other``` options - where it should say ```are depended on```.

Reply:
Request done in [0d6b6bb](https://github.com/squid-cache/squid/pull/999/commits/0d6b6bb4bd2601fdc13a3c28cd9be2834aba3714)

Reply:
> we have a documentation issue with finalizeConfig(). It clearly states that it is for `options that depend on other` options - where it should say `are depended on`.

FWIW, I do not know what documentation issue you are thinking about. Both "options that depend on X" and "options are depended on X" phrases mean the same thing to me, but the second phrase is grammatically [incorrect](https://forum.wordreference.com/threads/is-depended-on-vs-depends-on.3372884/post-17098859).